### ISSUE TITLE
Support for PlusROM functions

### ISSUE THREAD

**Issue by Al-Nafuur** [stella-emu/stella#592]
Support for PlusROM functions in Stella (if an internet connection is available).
more information about PlusROM functions:
- [PlusROM functions specs](http://pluscart.firmaplus.de/pico/?PlusROM)
- [PlusCart implementation](https://gitlab.com/firmaplus/atari-2600-pluscart)
- [PlusROM examples](https://gitlab.com/firmaplus/atari-2600-pluscart/-/tree/master/examples)
- [AtariAge thread about the PlusCart](https://atariage.com/forums/topic/297172-pluscart-an-inexpensive-diy-wifi-multicart/)
- [Javatari.js fork with PlusROM functions implementation](https://github.com/Firmaplus/javatari.js)

**Comment by sa666666** [comment]
I guess the first thing we'd need in Stella is the ability to communicate over a network.  Without that, there's nothing we can do with this.  So adding this feature may take some time.

**Comment by DirtyHairy** [comment]
I may look into adding this to 6502.ts / Stellerator though; will be much easier there.

**Comment by Al-Nafuur** [comment]
According to a suggestion of @thrust26 the specs for the PlusROM detection will change slightly.
We will use the interrupt vector to point to the hostname and path of the backend, so this information don't necessarily need to be encoded at the start of the ROM anymore.
This will make adding PlusROM functions to existing ROMs much easier, because the definition can be placed in an unused space of the ROM.

**Comment by sa666666** [comment]
Anyone have any idea what library we can use to get network functionality in Stella?  It would need to be GPLv2 compatible, and preferably C++ (although C is fine too).  And not too large.

Network functionality is eventually coming to C++ directly, but it's at least a year away.

This library should also make it possible to check for new versions of Stella, and eventually, perhaps for network play.

**Comment by sa666666** [comment]
This one looks good: https://think-async.com/Asio

**Comment by DirtyHairy** [comment]
> This one looks good: https://think-async.com/Asio

I think this is too low level. As far as I can see, we are only talking about HTTP, and for that, something like https://github.com/yhirose/cpp-httplib or even libcurl would be a better fit. Implementing HTTP properly on a low level network abstraction is a pain to get right, so I think it is better to reuse something that is already there.

If we want to support HTTPS, we'd also have to include OpenSSL into the game.

**Comment by DirtyHairy** [comment]
If I am not mistaken, ScummVM uses libcurl.

**Comment by thrust26** [comment]
I found this [website](https://superpowered.com/network-library-list).

libcurl looks most complete.

**Comment by DirtyHairy** [comment]
The only hunch I have with libcurl is that it is plain C and supports a whole bunch of protocols that are complete irrelevant for us. The library I linked above, on the other hand, is C++11,  reasonably small, header-only, and the code samples looks as minimal as I would expect an HTTP client to be.

**Comment by sa666666** [comment]
Easier is definitely the better way to go.  It may not be sufficient for network play, but as you say, that's so far away that we needn't worry about it now.

Would this cpp-httplib library allow for Stella to check for a new available version of Stella?  So we could at least kill two feature requests with one stone??

**Comment by DirtyHairy** [comment]
Yes. Github exposes an JSON/REST API that, among other things, allows to query the list of releases. However, that  runs via HTTPS, so we'd  need to add OpenSSL, which is quite  a monster.

**Comment by sa666666** [comment]
I wouldn't want to add SSL just for that one feature.  Much easier to just have a page on stella.github.io that we manually create, that contains the latest version number.

**Comment by DirtyHairy** [comment]
No dice, that's HTTPS, too --- HTTP is configured as a redirect to HTTPS.

**Comment by sa666666** [comment]
OK, I'll put it on my personal server :smiling_imp:

**Comment by DirtyHairy** [comment]
That will work fine 😛

**Comment by Al-Nafuur** [comment]
Gopher2600 (https://github.com/JetSetIlly/Gopher2600) has implemented the PlusROM functions too.
maybe a look at the source code can give some inspiration how to implement it:
https://github.com/JetSetIlly/Gopher2600/tree/master/hardware/memory/cartridge/plusrom

**Comment by thrust26** [comment]
@DirtyHairy @sa666666 IMO it is important to notify the user about the PlusROM communication result (positive or negative). Similar to the EEPROM writes for SaveKey/AtariVox. 

These are done using a callback function. Which would mean we have to change the constructors of all cart classes, no? Is there a better way?

**Comment by DirtyHairy** [comment]
There as an additional complication here: the notification would mean a call from the emulation thread into the UI code, which is illegal on some systems.

What we would need to do is add an "event" mechanism where the cart running on the emulation thread puts an event into a queue, and the main thread unqueues it and then reacts with a notification. That'd also solve the issue of having to modify the constructors.

**Comment by Al-Nafuur** [comment]
The user should be able to turn off these notifications.

**Comment by thrust26** [comment]
We could either provide `System` which will be most flexible or just the `EventHandler`. Or...?

**Comment by thrust26** [comment]
> The user should be able to turn off these notifications.

Sure. We already have such a setting for SaveKey writes.

**Comment by DirtyHairy** [comment]
> We could either provide System which will be most flexible or just the EventHandler. Or...?

I'd provide a queue as a singleton (or as a part of `OSystem`) that is guarded by a mutex and that holds lambdas. Each spin of the main loop empties the queue and executes the lambdas.

So, a cartridge (on the emulation thread) could queue a lambda that displays the notification, and the main loop (on the main thread) would unqueue and execute it.

**Comment by thrust26** [comment]
Sounds like a job for you. 😈

**Comment by thrust26** [comment]
@DirtyHairy @sa666666 I found a bug in PlusROM emulation. 

My code accidentally read from the PlusROM hotspots (there was data). The code still worked in Stella, but not on real hardware. 

Therefore we should react to peek from write hotspots too. And probably (still checking with Al_Nafuur) also on pokes to read hotspots.

**Comment by thrust26** [comment]
Regarding the messages: The functionality will be useful for SuperCharger, Compumate, KidVid etc. loads too.

**Comment by thrust26** [comment]
@sa666666 @DirtyHairy I think we can close this one now.

**Comment by Al-Nafuur** [comment]
@thrust26 
it looks like it can be changed from the command line too:
https://github.com/stella-emu/stella/blob/a3b34a8bd6be35ed196225e15f102ee1dae144f1/docs/index.html#L3291

**Comment by Al-Nafuur** [comment]
it should be Ok, if the commandline option is not a permanent change. Maybe just add a warning to the description of this option.

**Comment by thrust26** [comment]
It is currently permanent. Making the command line id temporary would be the best option.

**Comment by thrust26** [comment]
Why? I thought the "WE" was fixed? @DirtyHairy had added it to the original code too.

**Comment by Al-Nafuur** [comment]
It was fix at the "old" PlusStore-Id header to determine that it is a Emulator device, the new header has an "agent" field where the device is specified.

**Comment by thrust26** [comment]
> it should be Ok, if the commandline option is not a permanent change. Maybe just add a warning to the description of this option.

Why are you that much afraid of changing ids? I doubt many people would make any use of it.

**Comment by Al-Nafuur** [comment]
I am not that much afraid of it, but every "new" device-id will generate a new number and a new entry in the device list of the backends that have to keep track of it (e.g. the HSC):

![grafik](https://user-images.githubusercontent.com/15229085/138085400-65111318-a90c-42be-9332-e0ddadd05950.png)

**Comment by sa666666** [comment]
So this is completely done now?  I need to play around with this, to see how it works.  Thanks @DirtyHairy and @thrust26 for finishing this up.
